go/types/leaf_sequence.go

```diff
@@ -4,18 +4,139 @@
+)
+
 type leafSequence struct {
-	vrw    ValueReadWriter
-	length int
-	kind   NomsKind
+	vrw     ValueReadWriter
+	buff    []byte
+	offsets []uint32
+}
+
+const (
+	leafSequencePartKind   = 0
+	leafSequencePartLevel  = 1
+	leafSequencePartCount  = 2
+	leafSequencePartValues = 3
+)
+
+func newLeafSequence(kind NomsKind, count uint64, vrw ValueReadWriter, vs ...Value) leafSequence {
+	d.PanicIfTrue(vrw == nil)
+	w := newBinaryNomsWriter()
+	offsets := make([]uint32, len(vs)+leafSequencePartValues+1)
+	offsets[leafSequencePartKind] = 0
+	kind.writeTo(w)
+	offsets[leafSequencePartLevel] = w.offset
+	w.writeCount(0) // level
+	offsets[leafSequencePartCount] = w.offset
+	w.writeCount(count)
+	offsets[leafSequencePartValues] = w.offset
+	for i, v := range vs {
+		v.writeTo(w)
+		offsets[i+leafSequencePartValues+1] = w.offset
+	}
+	return leafSequence{vrw, w.data(), offsets}
+}
+
+// readLeafSequence reads the data provided by a decoder and moves the decoder forward.
+func readLeafSequence(dec *valueDecoder) leafSequence {
+	start := dec.pos()
+	offsets := skipLeafSequence(dec)
+	end := dec.pos()
+	return leafSequence{dec.vrw, dec.byteSlice(start, end), offsets}
+}
+
+func skipLeafSequence(dec *valueDecoder) []uint32 {
+	kindPos := dec.pos()
+	dec.skipKind()
+	levelPos := dec.pos()
+	dec.skipCount() // level
+	countPos := dec.pos()
+	count := dec.readCount()
+	offsets := make([]uint32, count+leafSequencePartValues+1)
+	offsets[leafSequencePartKind] = kindPos
+	offsets[leafSequencePartLevel] = levelPos
+	offsets[leafSequencePartCount] = countPos
+	offsets[leafSequencePartValues] = dec.pos()
+	for i := uint64(0); i < count; i++ {
+		dec.skipValue()
+		offsets[i+leafSequencePartValues+1] = dec.pos()
+	}
+	return offsets
+}
+
+func (seq leafSequence) decoder() *valueDecoder {
+	return newValueDecoder(seq.buff, seq.vrw)
+}
+
+func (seq leafSequence) decoderAtOffset(offset int) *valueDecoder {
+	return newValueDecoder(seq.buff[offset:], seq.vrw)
+}
+
+func (seq leafSequence) decoderAtPart(part uint32) *valueDecoder {
+	offset := seq.offsets[part] - seq.offsets[leafSequencePartKind]
+	return newValueDecoder(seq.buff[offset:], seq.vrw)
+}
+
+func (seq leafSequence) decoderSkipToValues() (*valueDecoder, uint64) {
+	dec := seq.decoderAtPart(leafSequencePartCount)
+	count := dec.readCount()
+	return dec, count
+}
+
+func (seq leafSequence) decoderSkipToIndex(idx int) *valueDecoder {
+	offset := seq.getItemOffset(idx)
+	return seq.decoderAtOffset(offset)
+}
+
+func (seq leafSequence) writeTo(w nomsWriter) {
+	w.writeRaw(seq.buff)
+}
+
+func (seq leafSequence) values() []Value {
+	dec, count := seq.decoderSkipToValues()
+	vs := make([]Value, count)
+	for i := uint64(0); i < count; i++ {
+		vs[i] = dec.readValue()
+	}
+	return vs
+}
+
+func (seq leafSequence) getCompareFnHelper(other leafSequence) compareFn {
+	dec := seq.decoder()
+	otherDec := other.decoder()
+
+	return func(idx, otherIdx int) bool {
+		offset := seq.getItemOffset(idx)
```

Comment:
again, it might be worth avoiding the two function invocations of getItemOffset by precomputing something that is closed over here.

Reply:
This one cannot be simplified more afaict